### PR TITLE
Fix several bugs in base64 library

### PR DESCRIPTION
**Pull request by salinecitrine**
Our base64 library previously had trouble decoding binary data, which is necessary for https://github.com/beyond-all-reason/Beyond-All-Reason/issues/3342 . This PR is the result of a discussion between myself and @p2004a (Marek found the issues, and I tested and implemented the changes).

See https://discord.com/channels/549281623154229250/1259438392316526602/1263522931657670738 for the discussion.

These changes were tested using randomized binary strings generated by a separate python program, and some were further verified by checking with https://base64.guru/converter/decode/hex. It has also been verified to work with at least one compressed json file for the Start Positions widget.

I think it is unlikely that this would negatively impact other uses; the fixes are both for edge cases that appear to only occur for binary data, which this library is rarely, if ever, used for.

I recommend reading and merging this PR as separate commits.

### CODE REVIEW

**Review by p2004a — APPROVED** [review]

**Comment by p2004a** [comment]
Note: merging as separate commits is disabled in this repo, only squash merge is allowed.

**Comment by salinecitrine** [comment]
> Note: merging as separate commits is disabled in this repo, only squash merge is allowed.

Ah, darn. Well, the formatting will be awkward I guess

File: common/luaUtilities/base64.lua [hunk 1]
```diff
@@ -5,58 +5,131 @@
...
+			(((#data - spos) > 1) and base64chars[lor(lsh(bytes[2] % 16, 2), rsh(bytes[3], 6))] or "=") ..
+			(((#data - spos) > 2) and base64chars[(bytes[3] % 64)] or "=")
 	end
 	result = table.concat(result)
 	return result
 end
 
 -- decoding table
-local base64bytes = {['A']=0,['B']=1,['C']=2,['D']=3,['E']=4,['F']=5,['G']=6,['H']=7,['I']=8,['J']=9,['K']=10,['L']=11,['M']=12,['N']=13,['O']=14,['P']=15,['Q']=16,['R']=17,['S']=18,['T']=19,['U']=20,['V']=21,['W']=22,['X']=23,['Y']=24,['Z']=25,['a']=26,['b']=27,['c']=28,['d']=29,['e']=30,['f']=31,['g']=32,['h']=33,['i']=34,['j']=35,['k']=36,['l']=37,['m']=38,['n']=39,['o']=40,['p']=41,['q']=42,['r']=43,['s']=44,['t']=45,['u']=46,['v']=47,['w']=48,['x']=49,['y']=50,['z']=51,['0']=52,['1']=53,['2']=54,['3']=55,['4']=56,['5']=57,['6']=58,['7']=59,['8']=60,['9']=61,['-']=62,['_']=63,['=']=nil}
+local base64bytes = {
```
Comment (p2004a) [inline common/luaUtilities/base64.lua:53]: Can we put it back in one line...? I just think spreading it like this... just makes me scroll more with 0 benefits...

Comment (salinecitrine) [reply common/luaUtilities/base64.lua:53]: I'll make them both a few lines.

**Pull request merged**